### PR TITLE
Draft

### PR DESCRIPTION
**Pull request by vmapetr**
# Description
New tool, Bug fixing, or Improvement?  
Please include a summary of the change and which issue is fixed. Also include relevant motivation and context.  
**For new tools, please provide total size and installation time.**

## Check list
- [ ] Related issue / work item is attached
- [ ] Tests are written (if applicable)
- [ ] Documentation is updated (if applicable)
- [ ] Changes are tested and related VM images are successfully generated

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -36,3 +36,29 @@ Here are a few things you can do that will increase the likelihood of your pull
...
+    └──/OS_TYPE
+        ├──/toolsets 
+        ├──/scripts
+        |   └──...
+        └──templates
+```
+- `docs` - ?
+- `help` - contains project documentation.
+- `helpers` - contains scripts related to image deployment.
+- `images.CI` - contains scripts that used in image-generation CI definitions. `/azure-pipelines` folder contains YAML defintions for CI pipelines.
```
Comment (AlenaSviridenko) [inline CONTRIBUTING.md]: that **are** used

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -36,3 +36,29 @@ Here are a few things you can do that will increase the likelihood of your pull
...
+- `scripts` - contains scripts with software installation logic.
```
Comment (AlenaSviridenko) [inline CONTRIBUTING.md]: I think not only the software installation logic is in this folder, but also image configuration and testing

File: docs/README.md [hunk 3]
```diff
@@ -1 +1,154 @@
-Placeholder for GitHub Pages
+## General
+The main purpose of Virtual-environments repository is to provide easy way to generate re-usable virtual images of required OS types using Azure infrastructure.
```
Comment (AlenaSviridenko) [inline docs/README.md]: I think this sentense is not needed, the purpose of this repo is pretty clear described in the main readme file

File: docs/README.md [hunk 4]
```diff
@@ -1 +1,154 @@
...
+Virtual-environments is using on [packer](https://www.packer.io/) tool and contain scripts with software installation logic for required Windows/Ubuntu tools.
```
Comment (AlenaSviridenko) [inline docs/README.md]: I would suggest to get rid of this section completely as it is not providing really useful information. 

This section can be replaced with a short announcement that this page contains information about prerequisites and image generation description.

File: docs/README.md [hunk 5]
```diff
@@ -1 +1,154 @@
...
+
+### Build Agent requirements
+To proceed with image generation, you need to prepare build agent with following requirements:
```
Comment (AlenaSviridenko) [inline docs/README.md]: this line can be removed too, title is pretty self-explanatory.

File: docs/README.md [hunk 6]
```diff
@@ -1 +1,154 @@
...
+- `OS` - Windows/Linux
+- `packer` - Can be downloaded from https://www.packer.io/downloads
+- `PowerShell 5.0 or higher` or `PSCore` for linux distributes.
+- `Azure CLI ` - https://docs.microsoft.com/ru-ru/cli/azure/install-azure-cli?view=azure-cli-latest
+
+### Azure subscription requirements
+#### How to generate service principal
+Packer will authorize in Azure infrastructure via Service Principal. If you want to prepare image-generation CI, or use packer manually, you need to create your SP with full read-write permissions for selected Azure subscription.
+[Check more details](https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-create-service-principal-portal)
```
Comment (AlenaSviridenko) [inline docs/README.md]: Check more details -> Detailed instruction can be found [here] (https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-create-service-principal-portal)

File: docs/README.md [hunk 7]
```diff
@@ -1 +1,154 @@
...
+
+### Image generation process
+Virtual environment project using packer tool, and contain list of templates for several OS types, such as Windows 2016/2019, Ubuntu 16.04/18.04/20.04.
```
Comment (AlenaSviridenko) [inline docs/README.md]: Project uses packer tool and contains a number of templates for several OS types: Windows 2016/2019, Ubuntu 16.04/18.04/20.04.

File: docs/README.md [hunk 8]
```diff
@@ -1 +1,154 @@
...
+Each image template contain set of scripts that will be executed on virtual machine.
```
Comment (AlenaSviridenko) [inline docs/README.md]: contain**s**

File: docs/README.md [hunk 9]
```diff
@@ -1 +1,154 @@
...
+Packer process will initialize connection to Azure subscription via Azure Cli, and deploy required resources, such as temporary resource group, network interfaces and VM from "clean" image that specified in template.
```
Comment (AlenaSviridenko) [inline docs/README.md]: CLI in uppercase. 
VM from **the** "clean" image that **is** specified in template.

File: docs/README.md [hunk 10]
```diff
@@ -1 +1,154 @@
...
+If deployment succeeded, build agent will connect with the deployed VM and begin to execute installation steps from selected template.
```
Comment (AlenaSviridenko) [inline docs/README.md]: will connect **to** the deployed VM
from **the** selected template.

File: docs/README.md [hunk 11]
```diff
@@ -1 +1,154 @@
...
+Dependend of template connection may be estableshed from SSH or WinRM protocols, which mean following ports between build agent, and temporary VM must be opened for connection.
```
Comment (AlenaSviridenko) [inline docs/README.md]: Connection can be established via SSH or WinRM, depending on image type, so the ports between build agent and temporary VM have to be opened for connection.

File: docs/README.md [hunk 12]
```diff
@@ -1 +1,154 @@
...
+If some of installation step from template fail, image generation will be aborted, and temporary VM will be terminated.
```
Comment (AlenaSviridenko) [inline docs/README.md]: If any template step fails, image generation will be aborted and temporary VM will be terminated.

File: docs/README.md [hunk 13]
```diff
@@ -1 +1,154 @@
...
+In case if all installation steps are succeeded, temporary VM snapshot will be converted to VHD image and will be downloaded into selected Azure Storage Account, from which it may be deployed further.
```
Comment (AlenaSviridenko) [inline docs/README.md]: If all steps are succeeded, a temporary VM snapshot will be converted to VHD and then uploaded to the specified Azure Storage Account. After that image can be deployed further from this account.

File: docs/README.md [hunk 14]
```diff
@@ -1 +1,154 @@
...
+
+### Azure DevOps selfhosted pool requirements
```
Comment (AlenaSviridenko) [inline docs/README.md]: self **-** hosted

File: docs/README.md [hunk 15]
```diff
@@ -1 +1,154 @@
...
+To connect with temporary VM packer will use WinRM, or SSH connections on public IP interfaces.
```
Comment (AlenaSviridenko) [inline docs/README.md]: To connect **to** temporary

File: docs/README.md [hunk 16]
```diff
@@ -1 +1,154 @@
...
+If you use self-hosted pool located in some Azure subscription, please make sure that HTTPS/SSH ports are allowed for incoming/outgoing connections.
+In case in firewall restrictions forbid connections via public addresses, you may deploy private virtual network resources, and pass them as arguments to packer, so temporary virtual machines can use private connections inside vlan.
```
Comment (AlenaSviridenko) [inline docs/README.md]: In case **if**

File: docs/README.md [hunk 17]
```diff
@@ -1 +1,154 @@
...
+If some of installation step from template fail, image generation will be aborted, and temporary VM will be terminated.
+In case if all installation steps are succeeded, temporary VM snapshot will be converted to VHD image and will be downloaded into selected Azure Storage Account, from which it may be deployed further.
+
+### Azure DevOps selfhosted pool requirements
+To connect with temporary VM packer will use WinRM, or SSH connections on public IP interfaces.
+If you use self-hosted pool located in some Azure subscription, please make sure that HTTPS/SSH ports are allowed for incoming/outgoing connections.
+In case in firewall restrictions forbid connections via public addresses, you may deploy private virtual network resources, and pass them as arguments to packer, so temporary virtual machines can use private connections inside vlan. 
+
+## Variables
+Each packer template include `variables` group that contains builder variables that used during image generation.
```
Comment (AlenaSviridenko) [inline docs/README.md]: include**s**
variables that used during image generation -> variables that are used in image generation process

File: docs/README.md [hunk 18]
```diff
@@ -1 +1,154 @@
...
+Builder variables can be passed to packer via predefined environment variables, or as direct arguments, in case if packer started manually.
+
+### Common
```
Comment (AlenaSviridenko) [inline docs/README.md]: look like this header can be removed

File: docs/README.md [hunk 19]
```diff
@@ -1 +1,154 @@
...
+
+### Required builder variables
+- `client_id` - The application ID of the AAD Service Principal. Requires client_secret.
```
Comment (AlenaSviridenko) [inline docs/README.md]: better put client_secret to code formatting `client_secret`

File: docs/README.md [hunk 20]
```diff
@@ -1 +1,154 @@
...
+- `object_id` - The object ID for the AAD SP. Optional, will be derived from the oAuth token if left empty.
```
Comment (AlenaSviridenko) [inline docs/README.md]: - `object_id` - The object ID for the AAD SP. Will be derived from the oAuth token if empty.

File: docs/README.md [hunk 21]
```diff
@@ -1 +1,154 @@
...
+- `client_secret` - A password/secret registered for the AAD SP.
+- `subscription_id` - The subscription to use.
+- `tenant_id` - The Active Directory tenant identifier with which your client_id and subscription_id are associated. If not specified, tenant_id will be looked up using subscription_id.
```
Comment (AlenaSviridenko) [inline docs/README.md]: I think it makes sense to put all variables names here to backticks to highlight them

File: docs/README.md [hunk 22]
```diff
@@ -1 +1,154 @@
...
+- `resource_group` - Resource group under which the final artifact will be stored.
+- `storage_account` - Storage account under which the final artifact will be stored.
+- `location` - Azure datacenter in which your VM will build.
```
Comment (AlenaSviridenko) [inline docs/README.md]: VM will **be** buil**t**

File: docs/README.md [hunk 23]
```diff
@@ -1 +1,154 @@
...
+If you use self-hosted pool located in some Azure subscription, please make sure that HTTPS/SSH ports are allowed for incoming/outgoing connections.
+In case in firewall restrictions forbid connections via public addresses, you may deploy private virtual network resources, and pass them as arguments to packer, so temporary virtual machines can use private connections inside vlan. 
+
+## Variables
+Each packer template include `variables` group that contains builder variables that used during image generation.
+Builder variables can be passed to packer via predefined environment variables, or as direct arguments, in case if packer started manually.
+
+### Common
+
+### Required builder variables
```
Comment (AlenaSviridenko) [inline docs/README.md]: A lot of variables in this section can be blank, so let's remove word "required" from the title

File: docs/README.md (hunk 22 above)
Comment (vmapetr) [reply docs/README.md]: This was actually, a copy-paste from packer documentation. Fixed.

**Pull request closed without merging**